Многозадачные запросы

Kot-Mars opened the pull request:
Jira - https://faang-school.atlassian.net/browse/BJS2-66597

PAIN((

Commit:
BJS2-66597: Create Entity
Create Table request_task
update table request

Kot-Mars committed:
BJS2-66557: Создание логики и тестов
Написаны тесты, создана бизнес логика и подключена к контролеру

SergeyErmoshenko commented on src/main/java/faang/school/accountservice/controller/RequestController.java:
```diff
@@ -39,4 +48,12 @@ public ResponseEntity<RequestResponseDto> updateRequest(
+    @GetMapping("/{id}")
+    public ResponseEntity<RequestStatusResponseDto> getStatus(@Valid @NotNull @PathVariable UUID id) {
+        Request request = requestRepository.findById(id).orElseThrow(() ->
+                new RequestNotFoundException(id.toString()));
```
А почему контроллер логику выполняет?

SergeyErmoshenko commented on src/main/java/faang/school/accountservice/repository/RequestRepository.java:
```diff
@@ -1,9 +1,13 @@
 import java.util.UUID;
 
 public interface RequestRepository extends JpaRepository<Request, UUID> {
+    List<Request> findAllByStatusAndScheduledAtBefore(RequestStatus status, LocalDateTime time);
```
Может лучше пагинацией доставать?

SergeyErmoshenko commented on src/main/java/faang/school/accountservice/service/scheduled/RequestSchedulerService.java:
```diff
@@ -0,0 +1,120 @@
+            scheduler.shutdownNow();
+
+            for (ThreadPoolExecutor executor : executorsByType.values()) {
+                executor.shutdownNow();
```
Если ты его закрыл его принудительно, то зачем потом добавлять проверку с ожиданием?

Kot-Mars committed:
BJS2-66557: Исправление ошибок
Перенес логику с контролера в сервис, добавил пагинацию, исправил executor.shutdown

PavelEzhkov commented on src/main/java/faang/school/accountservice/service/request/RequestExecutorServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+    @Transactional
+    public void execute(UUID requestId) {
+        Request request = getRequest(requestId);
+        request.setStatus(RequestStatus.IN_PROGRESS);
```
Нет проверки начального статуса запроса (может быть стоит проверять, что запрос в статусе PENDING)?

PavelEzhkov commented on src/main/java/faang/school/accountservice/service/request/RequestExecutorServiceImpl.java:
```diff
@@ -0,0 +1,69 @@
+                executor.executeTaskWithNewTx(request, task, handler);
+            } catch (Exception e) {
+                log.error(REQUEST_TASK_EXECUTION_FAILED, e);
+                throw new RequestExecutorException(REQUEST_TASK_EXECUTION_FAILED );
```
После ошибки запрос останется в статусе IN_PROGRESS

PavelEzhkov commented on src/main/java/faang/school/accountservice/service/request/RequestServiceImpl.java:
```diff
@@ -90,4 +92,13 @@ public RequestResponseDto updateRequest(RequestUpdateDto requestUpdateDto) {
+    public RequestStatusResponseDto getStatus(UUID id) {
+        Request request = requestRepository.findById(id)
+                .orElseThrow(() -> new RequestNotFoundException(id.toString()));
+        return new RequestStatusResponseDto(
```
А почему не через маппер?

PavelEzhkov commented on src/main/java/faang/school/accountservice/service/request/RequestTaskExecutor.java:
```diff
@@ -0,0 +1,49 @@
+    public void executeTaskWithNewTx(Request request, RequestTask task, RequestTaskHandler handler) {
+        try {
+            executeTask(request, task, handler);
+        } catch (Exception e) {
```
лучше ловить более конкретные типы исключений.

Kot-Mars committed:
BJS2-66597: Исправление ошибок
